pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -281,3 +308,67 @@ func convertToEnvoyJwtConfig(jwtRules []*v1beta1.JWTRule) *envoy_jwt.JwtAuthenti
 		Providers: providers,
 	}
 }
+
+// GetMutualTLSMode retuns the MutualTLSMode enum corresponding to the given peer authentiation policy.
+// If the input authentication is nil, returns MTLSPermissive.
+func GetMutualTLSMode(peer *v1beta1.PeerAuthentication) model.MutualTLSMode {
+	if peer == nil || peer.Mtls == nil {
+		return model.MTLSPermissive
+	}
+	switch peer.Mtls.Mode {
+	case v1beta1.PeerAuthentication_MutualTLS_DISABLE:
+		return model.MTLSDisable
+	case v1beta1.PeerAuthentication_MutualTLS_PERMISSIVE:
+		return model.MTLSPermissive
+	case v1beta1.PeerAuthentication_MutualTLS_STRICT:
+		return model.MTLSStrict
+	default:
+		// Should never reach here
+		log.Errorf("Unknown peer authentication %#v", peer)
+		return model.MTLSUnknown
+	}
+}
+
+// GetMostSpecificScopeConfig returns the config with the most specific scope, i.e config with
+// non-emtpy selector should be preferred over those without, one in non-root namespace should be
+// preferred over those in root namespace. Where there are more than one config in the same
+// scope, pick the one with stronger security (STRICT > PERMISSIVE > DISABLE).
+// The input configs is assumed to be a short list of configs that should be considerred for an
+// workload.
+func GetMostSpecificScopeConfig(rootNamespace string, configs []*model.Config) *model.Config {
+	// 0 - workload-specific config, define in the same namespace as workload (non-root namespace).
+	// 1 - workload-specific config, define in the root namespace.
```

Comment:
For  the first question, yes (for now), as we don't have the way to prevent selector used in root namespace yet (need some change in validation webhook to know what is the root namespace, and it may not be useful given root namespace can be changed).

For the second, both request authn and authz are appendable, so they don't need this logic to disambiguation.

Reply:
This is not a valid use case. WE dont respect workload specific config in the root namespace for sidecar or envoy filter. So lets follow the pattern for consistency and not create exceptions